pkg/ruler/prom_client.go

```diff
@@ -0,0 +1,124 @@
+package ruler
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/textproto"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log/level"
+	otgrpc "github.com/opentracing-contrib/go-grpc"
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/middleware"
+	"github.com/weaveworks/common/user"
+	"google.golang.org/grpc"
+
+	"github.com/cortexproject/cortex/pkg/util/grpcclient"
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+)
+
+const (
+	orgIDHeader      = "X-Scope-OrgID"
+	instantQueryPath = "/api/v1/query"
+	mimeTypeForm     = "application/x-www-form-urlencoded"
+	contentTypeJSON  = "application/json"
+)
+
+type promClient struct {
+	client               httpgrpc.HTTPClient
+	timeout              time.Duration
+	prometheusHTTPPrefix string
+	jsonDecoder          JsonDecoder
+}
+
+func NewPromClient(client httpgrpc.HTTPClient, timeout time.Duration, prometheusHTTPPrefix string) *promClient {
+	return &promClient{
+		client:               client,
+		timeout:              timeout,
+		prometheusHTTPPrefix: prometheusHTTPPrefix,
+		jsonDecoder:          JsonDecoder{},
+	}
+}
+
+func DialQueryFrontend(grpcClientConfig grpcclient.Config, queryFrontendUrl string) (httpgrpc.HTTPClient, error) {
```

Comment:
Should we use grpc client pool? https://github.com/cortexproject/cortex/blob/master/pkg/ring/client/pool.go#L60